src/main/java/net/minecraftforge/common/world/RegistryAccessExtension.java

```diff
@@ -0,0 +1,107 @@
+ * serialization/deserialization of the component type to and from json (or other data formats). Such
+ * components are server-side only.
+ * <p>
+ * The DeferredRegister held in this RegistryAccessExtension must be registered to the owning mod's
+ * event bus to function correctly.
+ *
+ * @param <T> The type of world-gen component.
+ */
+public class RegistryAccessExtension<T extends IForgeRegistryEntry<T>> extends ForgeRegistryEntry<RegistryAccessExtension<?>>
+{
+    private final ResourceKey<Registry<T>> registryKey;
+    private final Codec<T> elementCodec;
+    private final DeferredRegister<T> register;
+    private final Supplier<IForgeRegistry<T>> registry;
+
+    /**
+     * Creates a new RegistryAccessExtension with the given registry name, element type, and element Codec.
+     * This constructor automatically makes the backing Registry for the DeferredRegister.
+     *
+     * @param name  The registry name of the extension. The namespace of this ResourceLocation will be used as
+     *              the modid of the internally held DeferredRegister, and the path of the ResourceLocation will
+     *              define the filepath the data in datapacks.
+     * @param type  The registry entry type (must extend IForgeRegistryEntry).
+     * @param codec The codec for serializing/deserializing the entry to/from data formats (namely json).
+     */
+    public RegistryAccessExtension(ResourceLocation name, Class<T> type, Codec<T> codec)
+    {
+        setRegistryName(name);
+        this.registryKey = ResourceKey.createRegistryKey(name);
+        this.elementCodec = codec;
+        this.register = DeferredRegister.create(type, name.getNamespace());
+        this.registry = register.makeRegistry(name.getPath(), () -> new RegistryBuilder<T>().disableSaving().disableSync());
+    }
+
+    /**
+     * Creates a new RegistryAccessExtension from the given IForgeRegistry and element Codec.
+     * The registry name of the RegistryAccessExtension is derived from the name of the provided IForgeRegistry.
+     *
+     * @param registry The registry that elements will be registered to.
+     * @param codec    The codec for serializing/deserializing the entry to/from data formats (namely json).
+     */
+    public RegistryAccessExtension(IForgeRegistry<T> registry, Codec<T> codec)
+    {
+        setRegistryName(registry.getRegistryName());
+        this.registryKey = ResourceKey.createRegistryKey(registry.getRegistryName());
+        this.elementCodec = codec;
+        this.register = DeferredRegister.create(registry, registry.getRegistryName().getNamespace());
+        this.registry = Suppliers.memoize(() -> registry);
+    }
+
+    public Codec<T> getCodec()
+    {
+        return elementCodec;
+    }
+
+    public ResourceKey<? extends Registry<T>> getRegistryKey()
+    {
+        return registryKey;
+    }
+
+    public DeferredRegister<T> getRegister()
+    {
+        return register;
+    }
```

Comment:
The point of datapack types is to not have any internal object registries. We should not be giving users the object to add json objects via code.

Reply:
concur'd

Reply:
Builtin registry entries guarantee some baseline set of components that worldgen can be constructed with when there are no datapacks providing that content present

Reply:
That's just bad design on Mojang. The idea of a json based registry is that it can have no elements if so desired. No biomes would just be a void world, same with no generators. There can be some defaulted case sure, but that would be something non-changing that would be abstracted away within the builder of the component registry, not as an exposed built in element.

Reply:
Beyond providing a way to configure world-gen/dimensions etc I couldn't comment on Mojang's ideals behind the design of the datapack system. Absence of default values in the registries means that the behaviour where they would have been used has to be hardcoded instead which seems counter to the configurability aspect.

I guess forcing mods to ship with json definitions in their jar makes those more accessible to end-users which is nice. However, if they can't register in code though they can't data-gen them without having to write their own custom system for it

Reply:
Apologies, I'm not meaning to comment on Mojang's ideals, I'm just stating my own opinion on how I imagine it to be. 

As for registering in code for data generation, that is completely reasonable. However, that would be directly coupled to the data provider. We shouldn't allow the user to just register outside of that provider based context. But again, these are my opinions.

Reply:
Seems a bit arbitrary to me to be honest but appreciate you'll have more insight to the kinds of things that need protecting against with such restrictions